### DISCUSSION TITLE
Running Vue on Code-Server

### DISCUSSION THREAD

**Discussion by abhinav1912** [coder/code-server#3912] [Q&A] [answered]
I've gone through the documentation, but can't wrap my head around this. On starting a Vue server, it uses port 8080. How do I access it?

**Comment by jsjoeio** [comment]
Hmm I haven't worked with Vue in a long time, but code-server also uses the 8080 port. You should be able to change the port on Vue's side. Have you tried [this](https://reactgo.com/change-port-number-vue/)?

**Reply by abhinav1912** [reply]
Yes, changed it to 3000. ```ip:3000``` was inaccessible

**Reply by lg-nineoneninethreenine** [reply]
Try using 8988. Even though I don't have a Linux or any PC with Vue, try that and use this: `localhost:8988`. That has to work.

**Comment by jsjoeio** [comment, marked as answer]
I figured it out. It's similar to how we tell folks to [proxy Create React App](https://coder.com/docs/code-server/v3.11.0/guide#proxying-to-create-a-react-app)

1. add `vue.config.js`
2. Change `port` and `publicPath`
```js
module.exports = {
  devServer: {
    port: 3454,
  },
  publicPath: "/absproxy/3454",
}
```
3. access app at `<code-server-root>/absproxy/3454` e.g. `http://localhost:8080/absproxy/3454`

Read more about `publicPath` in the [Vue.js docs](https://cli.vuejs.org/config/#publicpath)


## Video

https://user-images.githubusercontent.com/3806031/128551223-e3758d79-c81e-44cf-9bd1-d9439a058ef7.mov

**Reply by abhinav1912** [reply]
Exactly what I needed. Thanks @jsjoeio!

**Reply by jsjoeio** [reply]
No problem!

**Reply by jsjoeio** [reply]
P.S. we will add this to the docs in https://github.com/cdr/code-server/issues/3918 so thanks for bringing it up!

**Reply by abhinav1912** [reply]
@jsjoeio just a follow-up question. Whatever changes I make aren't visible on the web app/page, even after reloading the Vue server and refreshing the page. However, opening the page in a new tab displays the correct content. Could this be related to the code-server setup, or a Vue fault?

**Reply by jsjoeio** [reply]
Hmmm 🤔 Good question. I believe Vue and other frontend apps use hot-module-reloading (HMR) and they probably have a WebSocket connection or something which listens for changes and then auto-reloads your app. My guess is Vue isn't talking to the right endpoint for your app, causing the changes to not be visible. 

@Blakeinstein did you ever figure this out?

Previous [thread](https://github.com/cdr/code-server/discussions/3836).

We may need to make changes in code-server but not fully certain. cc @code-asher

**Reply by Blakeinstein** [reply]
> even after reloading the Vue server and refreshing the page.

> @Blakeinstein did you ever figure this out?

Well you also need to tell the client to look for hmr on a different path other than root.
```js
module.exports = {
  devServer: {
    port: 3454,
    sockPath: 'sockjs-node'
  },
  publicPath: "/absproxy/3454",
}
```

should be all you need. If you are willing to give vite a shot, instead vue cli, for the vue template. You can read my setup guide here, https://github.com/Blakeinstein/codeserver-live#creating-a-template

**Reply by abhinav1912** [reply]
I'll try that out, but why does refresh not work and opening in a new tab does? It might be a caching issue, but that seems unlikely since this issue persists across browsers.

**Reply by Blakeinstein** [reply]
Well lets say you have the dev server on localhost:8080/absproxy/port

The devserver, expects the hmr socket to be on some path, usually the root, to listen to updates and tell the page to reload. Now since we have a proxied route. The socket isnt there at the expected path, which for our case the dev server usually thinks is localhost:8080 (we have code-server running on this, and internally the dev server socket is on localhost:port)

So we tell the dev server, to look for the socket on a specific route instead. Atleast that was my solution with vite, vue-cli uses webpack internally and I am pretty sure you should be able to replicate the setting here, as suggested by the docs https://coder.com/docs/code-server/v3.10.2/guide#proxying-to-create-a-react-app

basically along with the url you also set the path for the hmr socket.

**Reply by jsjoeio** [reply]
I knew you'd have the answer! Thanks for the help @Blakeinstein 🙌 (I'll also add the HMR part to the guide when I get this in the docs).

**Reply by bpmct** [reply]
Here's how I got Vue 3 working with an external domain FYI (e.g `code-server --link`):

```js
// vue.config.js
module.exports = {
    publicPath: (process.env.NODE_ENV=="development" ? "/absproxy/8080" : "/"),
    // options...
    devServer: {
        disableHostCheck: true,
        public: "https://bens_macbook-bpmct.cdr.co/absproxy/8080",
        sockPath: '/absproxy/8080/sockjs-node',
    }
}
```
